### PR TITLE
CUtlSymbolTableLarge and related things

### PR DESCRIPTION
**Pull request by vanz666**
Updated CUtlSymbolTableLarge and other important things that will be used in the future.

### CODE REVIEW

**Comment by GAMMACASE** [comment]
May I know where are all these changes came from?

**Comment by vanz666** [comment]
Reversing started from constructor and AddString function gradually exploring all related structures in server.dll/tier0.dll
You can find constructor by "HoldAnimation" and "CUtlSymbolTableLargeBase" (for ThreadSafe version)


cs2


AddString function can by found via "ERROR: CUtlSymbolTableLarge element limit of %u exceeded\n"
Testing was on Windows and there may be some other changes in the future.

**Comment by GAMMACASE** [comment]
I'll pinpoint a particular code lines where I have questions on where/how did you find/come up with them, and would like to get an answer on these

File: public/tier0/platform.h [hunk 1]
```diff
@@ -650,8 +657,14 @@ typedef unsigned int		uint;
...
-#define  stackfree( _p )			0
-
+#define stackalloc_aligned( _size, _align )		(void*)( ( ((uintp)alloca( ALIGN_VALUE( ( _size ) + (_align ),  ( _align ) ) )) + ( _align ) ) & ~( _align - 1 ) )
+
+// We should probably always just align to 16 bytes, stackalloc just causes too many problems without this behavior. Source2 does it already.
+// #define stackalloc( _size )							stackalloc_aligned( _size, 16 )
+
+#define  stackfree( _p )			0
+// two-argument ( type, #elements) stackalloc
+#define StackAlloc( typ, nelements ) ( ( typ * )	stackalloc_aligned( ( nelements ) * sizeof(typ), 16 ) )
```
Comment (GAMMACASE) [inline public/tier0/platform.h]: I'm interested in knowing how did you come up with these or what is required these to be setup like that? Specifically want to hear where did you take the comment from.

File: public/tier1/memblockallocator.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+#pragma once
+#endif
+
+#include "tier1/utlvector.h"
+
+typedef unsigned int MemBlockHandle_t;
+
+#define MEMBLOCKHANDLE_INVALID ((MemBlockHandle_t)~0)
+
+class CUtlMemoryBlockAllocator
```
Comment (GAMMACASE) [inline public/tier1/memblockallocator.h:27]: Where did you find this class definition? I'm aware that some methods from here are exported by tier0.dll, my question is where/how did you come up with the structure of that class?

File: public/tier0/threadtools.h [hunk 3]
```diff
@@ -595,22 +591,22 @@ class CInterlockedPtr
...
 
-	bool TryLock();
-	bool TryLock() const	{ return (const_cast<CThreadMutex *>(this))->TryLock(); }
+	bool TryLock( const char *pFileName, int nLine );
+	bool TryLock( const char *pFileName, int nLine ) const	{ return (const_cast<CThreadMutex *>(this))->TryLock( pFileName, nLine ); }
 
-	void LockSilent(); // A Lock() operation which never spews.  Required by the logging system to prevent badness.
-	void UnlockSilent(); // An Unlock() operation which never spews.  Required by the logging system to prevent badness.
+	void LockSilent( const char *pFileName, int nLine ); // A Lock() operation which never spews.  Required by the logging system to prevent badness.
+	void UnlockSilent( const char *pFileName, int nLine ); // An Unlock() operation which never spews.  Required by the logging system to prevent badness.
```
Comment (GAMMACASE) [inline public/tier0/threadtools.h:609]: I see that you've added the params in here, but I don't see them being used anywhere down the line? Can you explain why that change is required and how did you come up with it?

File: public/tier0/platform.h (hunk 1 above)
Comment (vanz666) [reply public/tier0/platform.h]: This code was taken from the leaked csgo source with one change (~_align to ~( _align - 1 ) in first line).
It is consistent with what is present in the cs2 assembly code. You can check this by compiling your program and checking the assembly code with what is in the cs2 code.

cs2-

File: public/tier1/memblockallocator.h (hunk 2 above)
Comment (vanz666) [reply public/tier1/memblockallocator.h:27]: A structure is a way of allocating memory in the form of fixed-size pages with blocks.
We can alloc blocks inside pages, and if the page size is not enough, we select a new page and place the block there.
The CUtlMemoryBlockAllocator::Alloc function returns a value in which the top 11 bits represent the page number (index in the m_MemPages vector) and the bottom 21 bits represent the offset to the block within the page.

CUtlMemoryBlockAllocator::CUtlMemoryBlockAllocator
cs2-mem1

CUtlMemoryBlockAllocator::Alloc
cs-mem2
cs2-mem3

File: public/tier0/threadtools.h (hunk 3 above)
Comment (vanz666) [reply public/tier0/threadtools.h:609]: These parameters may have been used in the debug build, or at least they are in the exported names. So I added them according to this prototype.

cs2-thread
cs2-thread2

Comment (GAMMACASE) [reply public/tier0/threadtools.h:609]: Which binary is that?

Comment (vanz666) [reply public/tier0/threadtools.h:609]: First screen from tier0.dll, second from server.dll

Comment (GAMMACASE) [reply public/tier0/threadtools.h:609]: Okay yeah, see them in the tier0.dll

File: public/tier0/platform.h (hunk 1 above)
Comment (GAMMACASE) [reply public/tier0/platform.h]: We don't support adding the leaked content to our codebase, can you please remove all the instances of the leaked code from the pr so we can move further with it?

File: public/tier1/generichash.h [hunk 4]
```diff
@@ -108,7 +108,15 @@ template<> inline unsigned HashItem<char *>(char * const &pszKey )
...
-uint32 MurmurHash2LowerCase( char const *pString, uint32 nSeed );
+uint32 MurmurHash2LowerCase( char const *pString, int nLength, uint32 nSeed )
+{
+	CBufferStringGrowable<128> buff;
+
+	buff.Insert( 0, pString, nLength );
+	buff.ToLowerFast( 0 );
+
+	return MurmurHash2( buff.Get(), nLength, nSeed );
+}
```
Comment (GAMMACASE) [inline public/tier1/generichash.h]: Can you move this function body to the generichash.cpp instead? And please replace this part of the code with ``MurMurHash2LowerCase`` then https://github.com/alliedmodders/hl2sdk/blob/3903ac9d50c522969603af1885b83a63ef81ae06/public/igameevents.h#L72-L75

**Review by GAMMACASE — CHANGES_REQUESTED** [review]

File: public/tier0/platform.h (hunk 1 above)
Comment (vanz666) [reply public/tier0/platform.h]: Removed.

File: public/tier1/generichash.h (hunk 4 above)
Comment (vanz666) [reply public/tier1/generichash.h]: Done.

Comment (GAMMACASE) [reply public/tier1/generichash.h]: Can you also replace the code I linked in the igameevents.h to use ``MurMurHash2LowerCase`` instead

**Review by psychonic — APPROVED** [review]

**Comment by GAMMACASE** [comment]
Thanks for the huge contribution, we appreciate your work!

**Pull request merged**